### ISSUE TITLE
Go 1.11 support.

### ISSUE THREAD

**Issue by dmitshur** [gopherjs/gopherjs#836]
This is an issue to track work to support Go 1.11. [`go1.11beta1`](https://groups.google.com/forum/#!topic/golang-announce/m1Szl8Mu_Fs) has been released today.

**Comment by shurcooL** [comment]
I've started working on this on [`go1.11`](https://github.com/gopherjs/gopherjs/compare/go1.11) branch. I've made good progress so far.

The current status is that most of the fundamental functionality is operational. Not all the tests are passing yet, but most are. The tests that are failing are mostly additions to Go 1.11 rather regressions in previous behavior. They're mostly new tests or bug fixes/improvements applied during 1.11 cycle that will need to be adapted to GopherJS to get tests to pass.

The current known failures on Linux ([from CI](https://circleci.com/gh/gopherjs/gopherjs/1618)) are:

- `crypto/des`, `TestEncryptTripleDES `: Error: crypto/des: invalid buffer overlap
- `crypto/rc4`, `TestGolden `: Error: crypto/rc4: invalid buffer overlap
- `go/doc`, `Test`: difference in output, there are trailing " = " in consts that don't have explicit values (this failure looks fascinating, I'm very curious to find out what's causing it)
- `reflect`, `TestTypes`: have `struct { int32 int32; int64 int64 }`, want `struct { int32; int64 }` (new testcase in Go 1.11, related to https://github.com/golang/go/commit/9558ba293f9ea1c9241eba99858eee449ba882d2 and https://github.com/golang/gofrontend/commit/023c3d4358d101c71ac1436065690eaec2ce138e)
- `github.com/gopherjs/gopherjs/tests`, `TestGoRepositoryCompilerTests `: 4 FAIL:
    - `fixedbugs/issue21221.go`: Error: fail
    - `fixedbugs/issue22662.go`: Error: got /tmp/issue22662.go.203426519:19015; want ??:1
    - `fixedbugs/issue22662b.go`: Error: runtime error: native function not implemented: syscall.runtime_BeforeFork (sounds like it'll need to be skipped due to use of unsupported `os.Exec`)
    - `fixedbugs/issue24547.go`: Error: FAIL

That work is up next. However, the compiler seems to be largely functional already.

**Comment by paralin** [comment]
For go1.11beta3 I am currently getting the following error (using vgo):

```
..\..\..\..\..\..\..\..\Go\src\runtime\error.go:44:18: invalid operation: e.concrete (variable of type *_type) has no field or method pkgpath
..\..\..\..\..\..\..\..\Go\src\runtime\error.go:44:42: invalid operation: e.asserted (variable of type *_type) has no field or method pkgpath
```

Thanks for the hard work!

**Comment by dmitshur** [comment]
@paralin You’ll need the patch from PR https://github.com/gopherjs/gopherjs/pull/846 to fix that. I haven’t had a chance to merge it yet, but the change looks right from a quick look.

**Comment by dmitshur** [comment]
> - `go/doc`, `Test`: difference in output, there are trailing " = " in consts that don't have explicit values (this failure looks fascinating, I'm very curious to find out what's causing it)

Investigating this now. Confirmed it's a regression (in GopherJS 1.11 WIP). Made a smaller reproduce:

```Go
package main

import (
    "fmt"
    "go/ast"
    "go/doc"
    "go/parser"
    "go/printer"
    "go/token"
    "os"
)

func main() {
    fset := token.NewFileSet()
    f, err := parser.ParseFile(fset, "issue16153.go", src, 0)
    if err != nil {
        panic(err)
    }
    apkg := &ast.Package{
        Name: "issue16153",
        Files: map[string]*ast.File{
            "issue16153.go": f,
        },
    }
    doc := doc.New(apkg, "foo/bar/issue16153", 0)
    err = printer.Fprint(os.Stdout, fset, doc.Consts[0].Decl)
    if err != nil {
        panic(err)
    }
    fmt.Println()
}

const src = `package issue16153
const (
    x2 uint8 = 255
    Y2
)
`
```

After printing the AST, I can see the difference that's causing this:

![image](https://user-images.githubusercontent.com/1924134/44313237-a7e40b80-a3d2-11e8-92b8-6f6e77aec1fa.png)

Might be a reflect bug, so it's great this got caught.

**Comment by dmitshur** [comment]
Investigating it further, I learned that the issue was due to changes to `go/doc` between 1.10 and 1.11, and not in GopherJS 1.11 WIP itself. The first `go/doc` commit where the problem occurs was https://github.com/golang/go/commit/c2366b27abe7e18b1b5642e5a9a009f5c41723ae#diff-23c81667acc0871b0276b715d353197f.

It had lines like these, which were very suspect:

```diff
@@ -158,6 +186,7 @@ func filterSpec(spec Spec, f Filter, export bool) bool {
 	switch s := spec.(type) {
 	case *ValueSpec:
 		s.Names = filterIdentList(s.Names, f)
+		s.Values = filterExprList(s.Values, f, export)
 		if len(s.Names) > 0 {
 			if export {
 				filterType(s.Type, f, export)
@@ -171,6 +199,7 @@ func (r *reader) filterSpec(spec ast.Spec) bool {
 		// always keep imports so we can collect them
 		return true
 	case *ast.ValueSpec:
+		s.Values = filterExprList(s.Values, ast.IsExported, true)
 		if len(s.Values) > 0 || s.Type == nil && len(s.Values) == 0 {
 			// If there are values declared on RHS, just replace the unexported
 			// identifiers on the LHS with underscore, so that it matches
```

After looking at the source of `filterExprList`, it was easy to guess that the GopherJS bug causing `go/doc` tests to fail in 1.11 was this subtle difference:

```Go
var s []int
s = s[0:0]
// s is []int(nil) in Go, but []int{} in GopherJS
```

Playground links confirming it:

- https://play.golang.org/p/JH-ofXXPQiW
- https://gopherjs.github.io/playground/#/JH-ofXXPQiW

(It's quite possible this is a reported issue in GopherJS, but hasn't been resolved yet.)

**Comment by dmitshur** [comment]
I've updated the `go1.11` branch to Go 1.11 RC 2, and resolved the few remaining blocking issues. [CI is now passing](https://circleci.com/gh/gopherjs/gopherjs/1636) on that branch.

The `go1.11` branch (PR #853) is now ready for review and testing. Please report any issues and opportunities to improve code. If there are no major issues discovered, the plan is to merge it when Go 1.11 is released.

**Comment by dmitshur** [comment]
GopherJS 1.11-1 is now released and available on `master` branch.

Blog post: https://medium.com/gopherjs/gopherjs-1-11-1-is-released-7aad12f7e646.

Tweet: https://twitter.com/GopherJS/status/1033473977509982208.

GopherJS Playground has been updated: https://gopherjs.github.io/playground/#/1pp8a59LAX.

Thanks everyone! 🎉 🏄‍♂️
